Algorithm.Framework/Portfolio/MeanVarianceOptimizationPortfolioConstruction.py

```diff
@@ -0,0 +1,136 @@
+# limitations under the License.
+
+from clr import AddReference
+AddReference("System")
+AddReference("QuantConnect.Algorithm")
+AddReference("QuantConnect.Common")
+AddReference("QuantConnect.Indicators")
+
+from System import *
+from QuantConnect import *
+from QuantConnect.Util import PythonUtil
+from QuantConnect.Algorithm import *
+from QuantConnect.Algorithm.Framework import *
+from QuantConnect.Algorithm.Framework.Alphas import *
+from QuantConnect.Data.UniverseSelection import *
+from datetime import timedelta
+import numpy as np
+import pandas as pd
+from scipy.optimize import minimize
+
+### <summary>
+### Provides an implementation of Mean-Variance portfolio optimization based on modern portfolio theory.
+### The target percent holdings of each security is calculated by maximizing the Sharpe ratio. 
+### The interval of weights in optimization method can be changed based on the long-short algorithm.
+### The default model uses the last three months daily price to calculate the optimal weight 
+### with the weight range from -1 to 1  
+### </summary>
+class MeanVarianceOptimizationPortfolioConstructionModel:
+    def __init__(self, resolution=Resolution.Daily, interval=63, min_weight=-1, max_weight=1):
+        """ 
+        Initialize the model
+        Args:
+            resolution:The resolution of the history price 
+            interval(int): The time interval of history price to calculate the weight 
+            min_weight(float): The lower bounds on portfolio weights
+            max_weight(float): The upper bounds on portfolio weights
+        """
+        self.securities = [ ]
+        self.resolution = resolution
+        self.interval = interval
+        self.min_weight = min_weight
+        self.max_weight = max_weight
+
+    def CreateTargets(self, algorithm, insights):
+        """ 
+        Create portfolio targets from the specified insights
+        Args:
+            algorithm:The algorithm instance
+            insights: The insights to create portoflio targets from
+        Returns: 
+            An enumerable of portoflio targets to be sent to the execution model
+        """
+        if len(self.securities) == 0:
+            return [ ]
+        
+        symbols = [insight.Symbol for insight in insights]
+        # request the daily history price of last year
+        hist = algorithm.History(symbols, self.interval, self.resolution)
+        df_price = hist['close'].unstack(level=0)
+        # calculate the daily return 
+        daily_return = (df_price / df_price.shift(1)).dropna()
+        weights = PortfolioOptimization(daily_return, self.min_weight, self.max_weight).opt_portfolio() 
+        weights.index = sorted([insight.Symbol.Value for insight in insights])
+        
+        # Create portfolio targets from the specified insights
+        for insight in insights:
+            yield PortfolioTarget.Percent(algorithm, insight.Symbol, insight.Direction * weights[insight.Symbol.Value])
```

Comment:
It's important to **never** index a dictionary using `Symbol.Value` as it's not guaranteed to be unique. Please just use the symbol object directly as the dictionary key.